### DISCUSSION TITLE
Native dlls in ClickOnce application

### DISCUSSION THREAD

**Discussion by erikture67** [QuestPDF/QuestPDF#992] [General]
I have a clickonce winforms application where I use QuestPDF. With the previous versions including skiasharp the native dlls was included when publishing the application. However since they have been replacede with QuestPdfs own skia dll I can not include them in the application. They do not show upp under application files. I think the reasion is that they are not marked as visible in the nuget spec for the targets of native files. 

Any suggestion how to include them in my application or do the nuget package have to change?

**Comment by erikture67** [comment]
I think that the cause of the problem is this:
The QuestPDT.targets file

<?xml version="1.0" encoding="utf-8"?>
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
    <ItemGroup>
        <None Include="$(MSBuildThisFileDirectory)\..\..\runtimes\**\*.*">
            <Visible>false</Visible>
            <Link>runtimes\%(RecursiveDir)%(Filename)%(Extension)</Link>
            <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
        </None>
    </ItemGroup>
</Project>

Changeing or removingthe visible false tag will make the QuestPdfSki.dlls visible in VisualStudio under ApplicationFiles when publishing ClickOnce applications.

I have made a temporary solution including the runtime dll in a nuget of my own this way and it works.

**Reply by MarcinZiabek** [reply]
Thank you for sharing this detail 😄  I have put it on the roadmap. Before introducing this change, I must investigate whether it negatively affects other users.

**Reply by erikture67** [reply]
Thanks!
I do not think that it can have any negative effect. The runtime dlls are copied to the bin folder when building the solution. The only problem is that they can note be included in a click once application. This is a "very common" problem. I have other third party products that I had the same problem and they did correct it.

**Reply by MarcinZiabek** [reply]
That's great! 😄

Is there any open-source project that you have already collaborated with regarding this type of issue? I am happy to use it as a guide. 

If everything goes well, one of the upcoming release candidate `2024.10.0` versions should have this problem fixed.

**Comment by erikture67** [comment]
Hello!

I had the same issue with Pdfium.Net.SDK and their pdfviewer.

/Erik

17 september 2024 kl. 18:24, "Marcin Ziąbek" ***@***.*** ***@***.******@***.***>)> skrev:
	That's great! 😄

	Is there any open-source project that you have already collaborated with regarding this type of issue? I am happy to use it as a guide.

	If everything goes well, one of the upcoming release candidate 2024.10.0 versions should have this problem fixed.

**Comment by erikture67** [comment]
Hello!

I also had it with ceTe.DynamicPDF.Viewer.NET and ceTe.DynamicPDF.Printing.NET

/Erik

17 september 2024 kl. 18:24, "Marcin Ziąbek" ***@***.*** ***@***.******@***.***>)> skrev:
	That's great! 😄

	Is there any open-source project that you have already collaborated with regarding this type of issue? I am happy to use it as a guide.

	If everything goes well, one of the upcoming release candidate 2024.10.0 versions should have this problem fixed.

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
